src/server/routes/oauth.ts

```diff
@@ -174,6 +235,73 @@ const authenticationHandler = async (
 			return redirectForGenericError(req, res);
 		}
 
+		const { user_groups, email_validated, sub, email } =
+			tokenSet.claims() as CustomClaims;
+
+		// We want to ensure that the user has a legacyIdentityId set in their
+		// Okta profile before carrying on. This is surfaced via the legacy_identity_id
+		// claim in the access token.
+		const refreshToken = tokenSet.refresh_token;
+		const { legacy_identity_id } = tokenSet.claims();
+		if (!refreshToken && !legacy_identity_id) {
+			// We can't do this step without the refresh token, so if it's missing, just continue
+			// to the callback function - we may encounter errors there.
+			trackMetric('OAuthAuthorization::ProvisioningFailure');
+			logger.error(
+				'Access token missing legacy_identity_id and no refresh token available',
+				undefined,
+				{
+					request_id: res.locals.requestId,
+				},
+			);
+		} else if (refreshToken && !legacy_identity_id) {
+			try {
+				// We have a refresh token and no legacy IDAPI ID, so we can try to fix this user.
+				const validTokenSet = await fetchValidTokenSet({
+					oktaId: sub,
+					email,
+					openIdClient,
+					tokenSet,
+					refreshToken,
+					ip: req.ip,
+					request_id: res.locals.requestId,
+				});
+				if (!validTokenSet) {
+					throw new Error('Failed to get a valid token set');
```

Comment:
why do we throw an exception in this case but not in the earlier case when there's no `legacy_identity_id` or `refreshToken` ?